framework/src/problems/FEProblemBase.C

```diff
@@ -933,7 +933,9 @@ FEProblemBase::prepare(const Elem * elem, THREAD_ID tid)
 
   _nl->prepare(tid);
   _aux->prepare(tid);
-  _assembly[tid]->prepare();
+  if (!_has_jacobian || !_const_jacobian)
+    _assembly[tid]->prepareJacobianBlock();
+  _assembly[tid]->prepareResidual();
   if (_has_nonlocal_coupling)
     _assembly[tid]->prepareNonlocal();
 
```

Comment:
I'm not sure what you mean about _displaced_problem ,  @fdkong .   I can try to do something, but i suspect it'll be easier for you to implement it rather than try to explain exactly what you want me to do!

Reply:
He is saying, that you need to make a similar change in `DisplacedProblem::prepare(const Elem * elem, THREAD_ID tid)`. Notice that that is called a few lines below this change. The reason is that if somebody runs with a displaced problem, there is a dense residual and jacobian block (like for the non-displaced one), and it would not be prepared and the code would crash with segfault like you saw it earlier. Note that you do not want to add `_has_jacobian` into `DisplacedProlblem`, you want to ask for it from `_mproblem` there (which is the corresponding `FEProblemBase` instance. You may need to add a getter for it. And it looks like you may need to add a getter for `_const_jacobian` as well. So the change in `DisplacedProblem` would look like:
```
- _assembly[tid]->prepare();
+ if (!_mproblem.hasJacobian() || !_mproblem.constJacobian())
+   _assembly[tid]->prepareJacobianBlock();
+ _assembly[tid]->prepareResidual();
```

Reply:
Done.  I'm still nervous playing with stuff i don't understand

Reply:
![this is fine](https://i.imgur.com/c4jt321.png)

Reply:
:-)